backend/hct_mis_api/one_time_scripts/program_cycle_data_migration.py

```diff
@@ -0,0 +1,175 @@
+                    [
+                        str(hh_id)
+                        for hh_id in Payment.objects.filter(parent__program_cycle=cycle).values_list(
+                            "household_id", flat=True
+                        )
+                    ]
+                )
+            hh_ids_in_cycles = hhs_in_cycles_dict[cycle_id_str]
+
+            # check any conflicts
+            if new_hh_ids.intersection(hh_ids_in_cycles):
+                continue
+
+            TargetPopulation.objects.filter(id=comparing_with_pp.target_population_id).update(program_cycle=cycle)
+            comparing_with_pp.program_cycle = cycle
+            hhs_in_cycles_dict[cycle_id_str].update(new_hh_ids)
+            break
+
+        if not comparing_with_pp.program_cycle:
+            cycle = create_new_program_cycle(
+                str(comparing_with_pp.program_id),
+                ProgramCycle.ACTIVE,
+                comparing_with_pp.start_date.date(),
+                comparing_with_pp.end_date.date(),
+            )
+            TargetPopulation.objects.filter(id=comparing_with_pp.target_population_id).update(program_cycle=cycle)
+            comparing_with_pp.program_cycle = cycle
+            hhs_in_cycles_dict[str(cycle.id)] = new_hh_ids
+
+        comparing_with_pp.save(update_fields=["program_cycle"])
+
+
+def program_cycle_data_migration() -> None:
+    start_time = timezone.now()
+    logger.info("Hi There! Started Program Cycle Data Migration.")
+    logger.info(f"Cycles before running creation: {ProgramCycle.objects.all().count()}")
+    for ba in BusinessArea.objects.all().only("id", "name"):
+        logger.info(f"Started processing {ba.name}...")
+
+        # FINISHED program
+        for finished_program in (
+            Program.objects.filter(business_area_id=ba.id)
+            .exclude(status__in=[Program.DRAFT, Program.ACTIVE])
+            .only("id", "name", "start_date", "end_date", "status")
+        ):
+            start_data = finished_program.start_date
+            end_data = finished_program.end_date
+            for pp in PaymentPlan.objects.filter(program_id=finished_program.id).only(
+                "id", "program_id", "target_population_id"
+            ):
+                processing_with_finished_program(pp, start_data, end_data)
+
+        # ACTIVE program
+        for program in (
+            Program.objects.filter(business_area_id=ba.id)
+            .exclude(status=Program.FINISHED)
+            .only("id", "name", "start_date", "end_date", "status")
+        ):
+            with transaction.atomic():
+                logger.info(f"** Creating Program Cycles for program {program.name}")
+
+                default_cycle = ProgramCycle.objects.filter(program_id=program.id).first()
+                if not default_cycle:
+                    logger.info(f"###### Default Program Cycles for program {program.name} does not exist")
+                    continue
```

Comment:
So when Program does not have any ProgramCycle now, we do not create any? Why is that? Or only DRAFT Programs do not have a cycle right now?

Reply:
going to fix this part... and have to check it any programs without cycles